python/ray/serve/_private/autoscaling_state.py

```diff
@@ -0,0 +1,335 @@
+import logging
+import time
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Set
+
+from ray.serve._private.autoscaling_policy import AutoscalingPolicyManager
+from ray.serve._private.common import (
+    DeploymentHandleSource,
+    DeploymentID,
+    ReplicaID,
+    TargetCapacityDirection,
+)
+from ray.serve._private.constants import (
+    RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE,
+    RAY_SERVE_MIN_HANDLE_METRICS_TIMEOUT_S,
+    SERVE_LOGGER_NAME,
+)
+from ray.serve._private.deployment_info import DeploymentInfo
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+@dataclass
+class HandleRequestMetric:
+    actor_id: str
+    handle_source: DeploymentHandleSource
+    queued_requests: float
+    running_requests: Dict[ReplicaID, float]
+    timestamp: float
+
+    @property
+    def total_requests(self) -> float:
+        return self.queued_requests + sum(self.running_requests.values())
+
+    @property
+    def is_serve_component_source(self) -> bool:
+        return self.handle_source in [
+            DeploymentHandleSource.PROXY,
+            DeploymentHandleSource.REPLICA,
+        ]
+
+
+class AutoscalingState:
+    """Manages autoscaling for a single deployment."""
+
+    def __init__(self, deployment_id: DeploymentID):
+        self._deployment_id = deployment_id
+
+        # Map from handle ID to (# requests recorded at handle, recording timestamp)
+        self._handle_requests: Dict[str, HandleRequestMetric] = dict()
+        self._requests_queued_at_handles: Dict[str, float] = dict()
+        # Number of ongoing requests reported by replicas
+        self._replica_average_ongoing_requests: Dict[str, float] = dict()
+
+        self._deployment_info = None
+        self._autoscaling_policy_manager = None
+        self._running_replicas: List[ReplicaID] = []
+        self._target_capacity: Optional[float] = None
+        self._target_capacity_direction: Optional[TargetCapacityDirection] = None
+
+    @property
+    def autoscaling_policy_manager(self) -> AutoscalingPolicyManager:
+        return self._autoscaling_policy_manager
```

Comment:
it seems like the `AutoscalingPolicyManager` was sort of already trying to accomplish what this PR is doing. Should we collapse the logic there into this class directly? Or am I misunderstanding the intended layering?

Reply:
Yup agreed, I think it'd be cleaner if we collapse them. I can do that in a followup PR? It will make this PR harder to review, will also need to change quite a few tests

Reply:
sure, good with that